### DISCUSSION TITLE
如何设置不解析 $(varname)

### DISCUSSION THREAD

**Discussion by guyutongxue** [xmake-io/xmake#2006] [Q&A] [answered]
我的项目中，要使用 `set_configvar` 和 `add_configfiles` 将一个文本文件嵌入到字符串中，方法大概是：

`config.h.in`

```cpp
constexpr const char* SCRIPT_SRC = R"(${SCRIPT_SRC})"
```

`xmake.lua`

```lua
    on_load(function (target)
         target:set("configvar", "SCRIPT_SRC", io.readfile("script.ps1"))
         target:add("configfiles", "config.h.in")
    end)
```

但如果 `script.ps1` 中包含 `$(...)` 的格式就会被意外替换掉。比如内容是

```powershell
$duration = $(Get-Time) - $begin
```

的话，`config.h` 中出现的就是 `$duration =  - $begin`。请问有什么好的解决办法吗？

**Comment by waruqi** [comment]
```
    add_configfiles("config.h.in", {pattern = "@(.-)@"})
```

切到其他匹配模式，文档有说明：https://xmake.io/#/zh-cn/manual/project_target?id=targetadd_configfiles

**Reply by guyutongxue** [reply]
这个我试过，不起作用。您这个匹配模式指的应该是将大括号变量替换 `${...}` 改为 `@...@` 模式，但这并不更改原先 `$(...)` [这种](https://xmake.io/#/zh-cn/manual/builtin_variables) 的替换规则

**Comment by waruqi** [comment, marked as answer]
这个改不了，这个是 set_configvar("", "$(projectdir)") 等内置变量转义，不属于 configfiles，是所有 set_/add_ 接口都会处理

你只能自己取出来，自己替换下 
```lua
        local var = io.readfile("/tmp/a")
        var = var:replace("$(", "%$(", {plain = true})
        target:set("configvar", "TEST", var)
```

**Reply by guyutongxue** [reply]
好的，谢谢您
